clash-protocols-memmap/src/Protocols/MemoryMap/Check.hs

```diff
@@ -0,0 +1,151 @@
+-- SPDX-FileCopyrightText: 2024 Google LLC
+--
+-- SPDX-License-Identifier: Apache-2.0
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | Validity checks performed on memory maps
+module Protocols.MemoryMap.Check where
+
+import Clash.Prelude
+
+import Protocols
+
+import Protocols.MemoryMap
+import Protocols.MemoryMap.Check.AbsAddress
+import Protocols.MemoryMap.Check.Overlap
+import Protocols.MemoryMap.FieldType (TypeName)
+import Protocols.MemoryMap.TypeCollect
+
+import GHC.Stack (SrcLoc (..))
+import Language.Haskell.TH hiding (location)
+import Text.Printf (printf)
+
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+
+{- | A validated memory map, containing a 'MemoryMap' with absolute addressess
+and the set of types used in device registers and their definitions.
+-}
+data MemoryMapValid = MemoryMapValid
+  { validMap :: MemoryMap
+  , validTypes :: Map.Map TypeName TypeDescription
+  }
+  deriving (Show)
+
+data MemoryMapValidationErrors = MemoryMapValidationErrors
+  { absAddrErrors :: [AbsAddressValidateError]
+  , overlapErrors :: [OverlapError]
+  }
+  deriving (Show)
+
+data CheckConfiguration = CheckConfiguration
+  { startAddr :: Address
+  , endAddr :: Address
+  }
+
+-- | Validate a 'MemoryMap' based on a configuration.
+check ::
+  CheckConfiguration -> MemoryMap -> Either MemoryMapValidationErrors MemoryMapValid
+check CheckConfiguration{..} mm@MemoryMap{..} =
```

Comment:
Could you use record dot syntax to extract the fields? It's pretty hard to see what comes from where.